File: Python/Longest Substring Without Repeating Characters.py
class Solution:
    """
    Time:   O(n)
    Memory: O(n)
    """

    def lengthOfLongestSubstring(self, string: str) -> int:
        last_index = {}
        start_ind = max_length = 0

        for ind, ch in enumerate(string):
            if ch in last_index and start_ind <= last_index[ch]:
                start_ind = last_index[ch] + 1

            last_index[ch] = ind
            max_length = max(max_length, ind - start_ind + 1)

        return max_length

File: Java/Longest Substring Without Repeating Characters.java
class Solution {

    /**
     * Time:   O(n)
     * Memory: O(n)
     */
    public int lengthOfLongestSubstring(String s) {
        int[] nextIndex = new int[128];
        int longest = 0, left = 0;

        for (int right = 0; right < s.length(); right++) {
            left = Math.max(nextIndex[s.charAt(right)], left);
            longest = Math.max(longest, right - left + 1);
            nextIndex[s.charAt(right)] = right + 1;
        }

        return longest;
    }
}
